source/gboardforensics/gatherers/trainingcache.d

```diff
@@ -0,0 +1,153 @@
+module gboardforensics.gatherers.trainingcache;
+
+import gboardforensics.gatherers;
+import gboardforensics.models.trainingcache;
+import gboardforensics.analysis : DB;
+
+import std.algorithm;
+import std.array;
+import std.conv;
+import std.file;
+import std.path;
+import std.range : ElementType;
+
+import core.thread;
+
+import d2sqlite3;
+
+class TrainingCacheGatherer : IGatherer
+{
+	this(string path)
+		in (path.exists())
+		in (path.isFile())
+	{
+		_trainingcache.path = path;
+		_db = cast(DB) path.baseName();
+	}
+
+	void gather()
+	{
+		Database db = Database(_trainingcache.path);
+		scope(exit) db.close();
+
+		switch (_db)
+		{
+			case DB.Trainingcache2: prepTrainingCache2(db); break;
+			default:
+		}
+	}
+
+	/**
+	Both 'trainingcache2.db' and 'trainingcache3.db' databases have almost
+	the same table structure. The only difference being the table
+	'training_input_events_table' which if it doesn't exists it is split in
+	two other tables 's_table' and 'tf_table'. The 's_table', according to
+	[this blog](http://www.swiftforensics.com/2021/01/gboard-has-some-interesting-data.html),
+	stores the information about the focused application and the 'tf_table'
+	contains information about all keystrokes. On the other hand
+	'training_input_events_table' has all this information in one place, however
+	keystrokes are not stored in plain text as it is in 'tf_table' but only in
+	the protobuf field.
+	*/
+	private void prepTrainingCache2(Database db)
+	{
+
+		// checks if the database contains the 'training_input_events_table'
+		immutable hasTIET = !db.execute("
+			SELECT name
+			FROM sqlite_master
+			WHERE LOWER(type) IN ('table','view')
+			AND LOWER(name) = 'training_input_events_table'"
+		).empty;
+
+		if (!hasTIET)
+		{
+			// inserted sequences
+			_trainingcache.inserted = db.execute("
+				SELECT datetime(_timestamp/1000, 'unixepoch') as time, f3 as sequence, _timestamp FROM tf_table
+			").map!((r) {
+				TrainingCache.Info info;
+				info.time = r["time"].as!string;
+				info.sequence = r["sequence"].as!string;
+				info.timestamp = r["_timestamp"].as!size_t;
+				return info;
+			}).array;
+
+			// deleted sequences
+			_trainingcache.deleted = db.execute("
+				SELECT datetime(_timestamp/1000, 'unixepoch') as time, f5 as sequence, _timestamp FROM d_table
+			").map!((r) {
+				TrainingCache.Info info;
+				info.time = r["time"].as!string;
+				info.sequence = r["sequence"].as!string;
+				info.timestamp = r["_timestamp"].as!size_t;
+				return info;
+			}).array;
+
+			// raw history
+			_trainingcache.rawHistory = db.execute("
+				SELECT time, sequence, deleted, _timestamp FROM (
+				SELECT * FROM (
```

Comment:
You can remove this duplication:
```sql
SELECT * FROM (
	SELECT datetime(_timestamp/1000, 'unixepoch') as time, f2, f3 as sequence, 'false' as deleted, _timestamp
	FROM tf_table
	UNION
	SELECT datetime(_timestamp/1000, 'unixepoch') as time, f2, f5 as sequence, 'true' as deleted, _timestamp
	FROM d_table
)
ORDER BY 1, 2
```

Move it into a `static immutable` with a meaningful name to keep it readable and use a the formatter here to append the subquery. Like `format!"SELECT * FROM (%s)"(FOO_QUERY)`

Reply:
done